src/storage/s3.rs

```diff
@@ -0,0 +1,244 @@
+use super::Blob;
+use failure::Error;
+use futures::Future;
+use postgres::Connection;
+use rusoto_core::region::Region;
+use rusoto_credential::DefaultCredentialsProvider;
+use rusoto_s3::{GetObjectRequest, PutObjectRequest, S3Client, S3};
+use std::convert::TryInto;
+use std::io::Read;
+use time::Timespec;
+
+pub(crate) struct S3Backend<'a> {
+    client: S3Client,
+    bucket: &'a str,
+}
+
+impl<'a> S3Backend<'a> {
+    pub(crate) fn new(client: S3Client, bucket: &'a str) -> Self {
+        Self { client, bucket }
+    }
+
+    pub(super) fn get(&self, path: &str) -> Result<Blob, Error> {
+        let res = self
+            .client
+            .get_object(GetObjectRequest {
+                bucket: self.bucket.to_string(),
+                key: path.into(),
+                ..Default::default()
+            })
+            .sync()?;
+
+        let mut b = res.body.unwrap().into_blocking_read();
+        let mut content = Vec::with_capacity(
+            res.content_length
+                .and_then(|l| l.try_into().ok())
+                .unwrap_or(0),
+        );
+        b.read_to_end(&mut content).unwrap();
+
+        let date_updated = parse_timespec(&res.last_modified.unwrap())?;
+
+        Ok(Blob {
+            path: path.into(),
+            mime: res.content_type.unwrap(),
+            date_updated,
+            content,
+        })
+    }
+
+    pub(super) fn store_batch(&self, batch: &[Blob]) -> Result<(), Error> {
+        let mut rt = tokio::runtime::Runtime::new().unwrap();
+        let mut attempts = 0;
+
+        loop {
+            let mut futures = Vec::new();
```

Comment:
Can pre-allocate this
```suggestion
            let mut futures = Vec::with_capacity(batch.len());
```